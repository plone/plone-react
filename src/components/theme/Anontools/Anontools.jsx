/**
 * Anontools component.
 * @module components/theme/Anontools/Anontools
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { List } from 'semantic-ui-react';
import { FormattedMessage } from 'react-intl';

import { settings } from '~/config';

/**
 * Anontools container class.
 * @class Anontools
 * @extends Component
 */
class Anontools extends Component {
  /**
   * Property types.
   * @property {Object} propTypes Property types.
   * @static
   */
  static propTypes = {
    token: PropTypes.string,
    content: PropTypes.shape({
      '@id': PropTypes.string,
    }),
  };

  /**
   * Default properties.
   * @property {Object} defaultProps Default properties.
   * @static
   */
  static defaultProps = {
    token: null,
    content: {
      '@id': null,
    },
  };

  /**
   * Render method.
   * @method render
   * @returns {string} Markup for the component.
   */
  render() {
    return (
      !this.props.token && (
        <List floated="right" horizontal>
          {/* needs divs around links for a11y, and semanticui insists on using List as div role="list" instead of simply using <ul></ul> */}
          <div role="listitem" className="item">
            <Link
              to={`/login${
                this.props.content
                  ? `?return_url=${this.props.content['@id'].replace(
                      settings.apiPath,
                      '',
                    )}`
                  : ''
              }`}
            >
              <FormattedMessage id="Log in" defaultMessage="Log in" />
            </Link>
          </div>
          <div role="listitem" className="item">
            <Link to="/register">
              <FormattedMessage id="Register" defaultMessage="Register" />
            </Link>
          </div>
        </List>
      )
    );
  }
}

export default connect(state => ({
  token: state.userSession.token,
  content: state.content.data,
}))(Anontools);
