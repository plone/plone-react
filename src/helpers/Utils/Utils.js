import React from 'react';
import { transform, isEqual, isObject } from 'lodash';

import loadable from '@loadable/component';

const LoadableToast = loadable.lib(() => import('react-toastify'));

/**
 * Deep diff between two object, using lodash
 * @param  {Object} object Object compared
 * @param  {Object} base   Object to compare with
 * @return {Object}        Return a new object who represent the diff
 */
export function difference(object, base) {
  return transform(object, (result, value, key) => {
    if (!isEqual(value, base[key])) {
      result[key] =
        isObject(value) && isObject(base[key])
          ? difference(value, base[key])
          : value;
    }
  });
}

/**
 * Throw an error if the wrapped function returns undefined
 *
 * @param {Function} func
 */
export const safeWrapper = (func) => (config) => {
  const res = func(config);
  if (typeof res === 'undefined') {
    throw new Error(`Configuration function doesn't return config, ${func}`);
  }
  return res;
};

/**
 * A helper to pipe a configuration object through configuration loaders
 *
 * @param {Array} configMethods A list of configuration methods
 * @param {Object} config The Volto singleton config object
 */
export function applyConfig(configMethods, config) {
  return configMethods.reduce((acc, apply) => safeWrapper(apply)(acc), config);
}

/**
 * A HOC factory that propagates the status of asyncConnected requests back to
 * the main server process, to allow properly expressing an error status as
 * HTTP status code
 *
 * @param {} code HTTP return code
 */
export function withServerErrorCode(code) {
  return (WrappedComponent) => (props) => {
    if (props.staticContext && Object.keys(props.staticContext).length === 0) {
      const { staticContext } = props;
      staticContext.error_code = code;
      staticContext.error = props.error;
    }
    return <WrappedComponent {...props} />;
  };
}

/**
 * withToastify. A HOC that lazy-loads react-toastify and injects it as
 * the `toastify` property to the wrapped component
 *
 * @param {} WrappedComponent
 * @return Component
 */
export function withToastify(WrappedComponent) {
  class WithLoadableToastify extends React.Component {
    toastify = React.createRef();

    render() {
      return (
        <>
          <LoadableToast ref={this.toastify} />
          <WrappedComponent
            {...this.props}
            toastify={
              this.toastify.current ? this.toastify.current.toast : null
            }
          />
        </>
      );
    }
  }

  WithLoadableToastify.displayName = `WithLoadableToastify(${getDisplayName(
    WrappedComponent,
  )})`;
  return WithLoadableToastify;
}

function getDisplayName(WrappedComponent) {
  return WrappedComponent.displayName || WrappedComponent.name || 'Component';
}
