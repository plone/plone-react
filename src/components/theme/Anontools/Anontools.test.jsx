import React from 'react';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-intl-redux';
import { MemoryRouter } from 'react-router-dom';

import Anontools from './Anontools';
import { arrayWIdsToObject } from '@plone/volto/helpers/Utils/Utils';

const mockStore = configureStore();

const actions = { user: [{ id: 'login' }] };
const actionsById = arrayWIdsToObject(actions);

describe('Anontools', () => {
  it('renders an anontools component when no use is logged in', () => {
    const store = mockStore({
      actions: { actions, actionsById },
      content: { data: { '@id': 'myid' } },
      intl: {
        locale: 'en',
        messages: {},
      },
    });
    const component = renderer.create(
      <Provider store={store}>
        <MemoryRouter>
          <Anontools />
        </MemoryRouter>
      </Provider>,
    );
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('should not render an anontools component when a user is logged in', () => {
    const actions = { user: [{ id: 'logout' }] };
    const store = mockStore({
      actions: { actions, actionsById: arrayWIdsToObject(actions) },
      content: { data: {} },
      intl: {
        locale: 'en',
        messages: {},
      },
    });
    const component = renderer.create(
      <Provider store={store}>
        <MemoryRouter>
          <Anontools />
        </MemoryRouter>
      </Provider>,
    );
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });
});
