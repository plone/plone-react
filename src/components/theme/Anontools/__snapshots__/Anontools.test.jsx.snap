// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Anontools renders an anontools component when no token is specified 1`] = `
<div
  className="ui horizontal right floated list"
  role="list"
>
  <div
    className="item"
    role="listitem"
  >
    <a
      href="/login?return_url=myid"
      onClick={[Function]}
    >
      Log in
    </a>
  </div>
  <div
    className="item"
    role="listitem"
  >
    <a
      href="/register"
      onClick={[Function]}
    >
      Register
    </a>
  </div>
</div>
`;

exports[`Anontools should not render an anontools component when a token is specified 1`] = `null`;
